docs/rfc/rfc-027-p2p-message-bandwidth-report.md

```diff
@@ -0,0 +1,137 @@
+The consensus reactor starts a new `gossipDataRoutine` for each peer it connects to.
+This routine repeatedly picks a part of the block that Tendermint believes the peer does not know about yet and gossips it to the peer.
+The set of `BlockParts` that Tendermint considers its peer as having is only updated in one of four ways:
+
+1. [We receive a block part from the peer][block-part-receive].
+2. [We send][block-part-send-1] [the peer a block part][block-part-send-2].
+3. Our peer tells us about the parts they have block [via `NewValidBlock` messages][new-valid-block-message-send]. This message is only sent when the validator has a quorum of prevotes or precommits for a block.
+4. Our peer tells us they have entered a new round [via a `NewRoundStep` message][new-round-step-message-send].
+
+Each node receives block parts from all of its peers. The particular block part to send at any given time is randomly selected from the set of parts that the peer node is not yet known to have. Given that these are the only times that Tendermint learns of its peers' block parts, it's very likely that a node has an incomplete understanding of its peers' block parts and is transmitting block parts to a peer that the peer has received from some other node.
+
+Multiple potential mechanisms exist to reduce the number of duplicate block parts a node receives. One set of mechanisms relies on more frequently communicating the set of block parts a node needs to its peers. Another potential mechanism requires a larger overhaul to the way blocks are gossiped in the network.
+
+
+#### Mempool Tx Transmission
+
+The Tendermint mempool stages transactions that are yet to be committed to the blockchain and communicates these transactions to its peers. Each message contains one transaction. Data collected from the Blockpane node running on Osmosis indicates that the validator sent about 12 gigabytes of `Txs` messages during the nearly 3 hour observation period.
+
+The Tendermint mempool starts a new [broadcastTxRoutine][broadcast-tx-routine] for each peer that it is informed of. The routine sends all transactions that the mempool is aware of to all peers with few exceptions. The only exception is if the mempool received a transaction from a peer, then it marks it as such and won't resend to that peer. Otherwise, it retains no information about which transactions it already to sent to a peer. In some cases it may therefore resend transactions the peer already has. This can occur if the mempool removes a transaction from the `CList` data structure used to store the list of transaction while it is about to be sent and if the transaction was the tail of the `CList` during removal. This will be more likely to occur if a large number of transactions from the end of the list are removed during `RecheckTx`, since multiple transactions will become the tail and then be deleted. It is unclear at the moment how frequently this occurs on production chains.
```

Comment:
```suggestion
The Tendermint mempool starts a new [broadcastTxRoutine][broadcast-tx-routine] for each peer that it is informed of. The routine sends all transactions that the mempool is aware of to all peers with few exceptions. The only exception is if the mempool received a transaction from a peer, then it marks it as such and won't resend to that peer. Otherwise, it retains no information about which transactions it already sent to a peer. In some cases it may therefore resend transactions the peer already has. This can occur if the mempool removes a transaction from the `CList` data structure used to store the list of transactions while it is about to be sent and if the transaction was the tail of the `CList` during removal. This will be more likely to occur if a large number of transactions from the end of the list are removed during `RecheckTx`, since multiple transactions will become the tail and then be deleted. It is unclear at the moment how frequently this occurs on production chains.
```